frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.jsx

```diff
@@ -29,10 +31,12 @@ export default class ExpressionWidget extends Component {
   }
 
   UNSAFE_componentWillReceiveProps(newProps) {
-    this.setState({
-      name: newProps.name,
-      expression: newProps.expression,
-    });
+    if (!this.state || !_.isEqual(this.props.expression, newProps.expression)) {
```

Comment:
All kudos to @ariya here…he did it.

Reply:
Wow, I'm just curious why don't we define `state` as a class property right away? This isn't a common approach, maybe it's worth a comment if it can't be avoided?

Reply:
The console is littered with lifecycle warnings. We prefixed them with `UNSAFE_` but this was just a temporary fix (that doesn't work in strict mode, if I understand correctly).
Just wanted to ask does it make sense to follow up with the PR to get rid of the unsupported lifecycle method and do it properly (according to the guidelines from the warning):

> * Move data fetching code or side effects to componentDidUpdate.
> * If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
> * Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Reply:
@kulyk @nemanjaglumac  So far I've been quite confused but this (and other things).

To be honest, it's been years since I used Class components at all in React.

I'd be quite inclined to support improvements around this, at the same time it's a bit early for me to have a solid opinion.

But definitely let's keep talking about it, please.

Reply:
Sorry, I probably asked a not really clear question. I wanted to know why do we check for `!this.state`? This check means that the state is not initialized in the class constructor. Why so if we can e.g. define a state as usual:

```jsx
class TheComponent extends React.Component {
    state = {
        name: "some-initial-value",
        expression: "some-initial-value",
    }
}

Reply:
@kulyk It's used in some other places (likely legacy), see also ExpressionEditorTextfield.

While this leaves a great deal to be desired, to be honest I think we shall just do the minimum here (just like what works in ExpressionEditorTextfield) and perform a better refactoring/improvement in the master branch instead.

Reply:
@kulyk, imo it would be extremely helpful if you created a Github issue with what you propose to improve here.

Could be labeled as `.TechDebt`.

Even if we don't get around to it very soon, giving these issues visibility should help us, as a team, to elect priorities in improving the codebase.

cc @ariya @nemanjaglumac